File: packages/guards/src/semantic-similarity-check.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface SemanticSimilarityCheckOptions {
  action: 'block' | 'warn';
  threshold?: number;
}

function getTrigrams(text: string): Set<string> {
  const s = text.toLowerCase().trim();
  const trigrams = new Set<string>();
  for (let i = 0; i <= s.length - 3; i++) {
    trigrams.add(s.slice(i, i + 3));
  }
  return trigrams;
}

function trigramSimilarity(a: string, b: string): number {
  const tA = getTrigrams(a);
  const tB = getTrigrams(b);
  if (tA.size === 0 && tB.size === 0) return 1;
  if (tA.size === 0 || tB.size === 0) return 0;
  let intersection = 0;
  for (const t of tA) {
    if (tB.has(t)) intersection++;
  }
  return intersection / Math.max(tA.size, tB.size);
}

export function semanticSimilarityCheck(options: SemanticSimilarityCheckOptions): Guard {
  const threshold = options.threshold ?? 0.8;

  return {
    name: 'semantic-similarity-check',
    version: '0.1.0',
    description: 'Checks if output is too similar to input (potential echo/copy)',
    category: 'ai',
    supportedStages: ['output'],
    async check(text: string, ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const input = (ctx as unknown as Record<string, unknown>).inputText as string | undefined;

      if (!input) {
        return {
          guardName: 'semantic-similarity-check',
          passed: true,
          action: 'allow',
          score: 0,
          latencyMs: Math.round(performance.now() - start),
        };
      }

      const similarity = trigramSimilarity(input, text);
      const triggered = similarity >= threshold;

      return {
        guardName: 'semantic-similarity-check',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score: similarity,
        latencyMs: Math.round(performance.now() - start),
        details: { similarity, threshold },
      };
    },
  };
}

File: python/open_guardrail/guards/semantic_similarity_check.py
"""Detect echo or copy via trigram similarity."""
from __future__ import annotations

import re
import time

from open_guardrail.core import GuardResult


def _trigrams(text: str) -> set:
    clean = re.sub(r"\s+", " ", text.lower().strip())
    if len(clean) < 3:
        return set()
    return {clean[i : i + 3] for i in range(len(clean) - 2)}


class _SemanticSimilarityCheck:
    def __init__(
        self,
        *,
        action: str = "warn",
        threshold: float = 0.8,
    ) -> None:
        self.name = "semantic-similarity-check"
        self.action = action
        self.threshold = threshold

    def check(
        self,
        text: str,
        stage: str = "output",
        *,
        reference: str = "",
    ) -> GuardResult:
        start = time.perf_counter()

        t1 = _trigrams(reference)
        t2 = _trigrams(text)

        if not t1 or not t2:
            similarity = 0.0
        else:
            intersection = t1 & t2
            union = t1 | t2
            similarity = len(intersection) / len(union)

        triggered = similarity >= self.threshold
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="semantic-similarity-check",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=round(similarity, 4),
            message=(
                "Response too similar to input"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {
                    "similarity": round(similarity, 4),
                    "threshold": self.threshold,
                    "reason": (
                        "Response appears to echo or"
                        " copy the input text"
                    ),
                }
                if triggered
                else None
            ),
        )


def semantic_similarity_check(
    *,
    action: str = "warn",
    threshold: float = 0.8,
) -> _SemanticSimilarityCheck:
    return _SemanticSimilarityCheck(
        action=action, threshold=threshold
    )
